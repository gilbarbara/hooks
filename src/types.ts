import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
