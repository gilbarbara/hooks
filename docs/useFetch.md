# useFetch
A custom hook for making HTTP requests using the Fetch API.

Supports caching, dynamic URLs, retries (exponential or linear backoff), error handling, and flexible configuration options.  
It provides helper methods to check the request's status and allows manual or automatic execution.

## Usage

```tsx
import { useFetch, USE_FETCH_STATUS } from '@gilbarbara/hooks';

interface ToDo {
  completed: boolean;
  id: number,
  title: string;
  userId: number;
}

function Component() {
  const { data, error, isError, isSuccess, refetch, status } = useFetch<Array<ToDo>>(
    'https://jsonplaceholder.typicode.com/todos'
  );

  // Trigger fetch when needed
  const handleRefetch = async () => {
    await refetch();
  };

  return (
    <div>
      {status === USE_FETCH_STATUS.LOADING && <p>Loading</p>}
      {isError() && <p>{error?.toString()}</p>}
      {isSuccess() && (
        <>
          <button onClick={handleRefetch}>Refetch</button>
          <pre>{JSON.stringify(data.slice(0, 10), null, 2)}</pre>
        </>
      )}
    </div>
  );
}
```

**Advanced Usage with Manual Trigger**

```tsx
import { useState } from 'react';
import { useFetch } from '@gilbarbara/hooks';

interface ToDo {
  completed: boolean;
  id: number,
  title: string;
  userId: number;
}

function Component() {
  const [wait, setWait] = useState(true);

  const { error, data, isError, isLoading, isPaused, isSuccess, refetch } = useFetch<Array<ToDo>>({
    url: 'https://jsonplaceholder.typicode.com/todos',
    retry: 3, // Retry up to 3 times on failure
    retryDelay: (attempt: number) => Math.min(2 ** attempt * 1000, 30 * 1000), // Exponential backoff
    wait,
  });

  const handleFetch = () => {
    setWait(false);
  };

  const handleRefetch = () => {
    refetch();
  };

  return (
    <div>
      <button onClick={handleFetch} disabled={isLoading() || isSuccess()}>
        Fetch Data
      </button>
      {isLoading() && <p>Loading...</p>}
      {isError() && <p>{error?.toString()}</p>}
      {isSuccess() && (
        <>
          {data?.slice(0, 10).map(user => (
            <p>{user.title}</p>
          ))}
          <button onClick={handleRefetch}>Refetch</button>
        </>
      )}
    </div>
  );
}
```

## Caching

The useFetchCache utility provides advanced control over the global cache used by useFetch. You can use it to preload, clear, or inspect cached data. This utility is optional and meant for scenarios requiring fine-grained cache management.

**Example**

```tsx
import { useFetch, useFetchCache } from '@gilbarbara/hooks';

// Preload data with a 1-minute TTL
useFetchCache.set('https://api.example.com/data', { preloaded: true }, 60000);

function MyComponent() {
  const { data, isCached, isLoading, isSuccess } = useFetch('https://api.example.com/data', {
    cacheTTL: 60000, // Enable caching
  });

  return (
    <div>
      {isLoading() && <p>Loading...</p>}
      {isSuccess() && (
        <p>
          {isCached ? 'Loaded from cache.' : 'Fetched from server.'}
          <pre>{JSON.stringify(data, null, 2)}</pre>
        </p>
      )}
    </div>
  );
}
```

## Reference

```typescript

const USE_FETCH_STATUS = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
} as const;

type UseFetchStatus = keyof typeof USE_FETCH_STATUS;

interface UseFetchOptions {
  body?: BodyInit | Record<string, any>;
  /**
   * Time to cache the request if provided.
   * When the cache is expired, the request will be triggered again.
   */
  cacheTTL?: number;
  headers?: Record<string, string>;
  /**
   * HTTP method.
   * @default: 'GET'
   */
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'HEAD' | 'OPTIONS' | 'CONNECT' | 'TRACE';
  /**
   * Request mode.
   * @default: 'cors'
   */
  mode?: 'cors' | 'navigate' | 'no-cors' | 'same-origin';
  /**
   * Number of retries.
   * @default: 3
   */
  retry?: number | false;
  /**
   * Time to wait before retrying.
   * A function like attempt => Math.min(attempt > 1 ? 2 ** attempt * 1000 : 1000, 30 * 1000) applies exponential backoff.
   * A function like attempt => attempt * 1000 applies linear backoff.
   * @default: attempt => attempt * 1000
   */
  retryDelay?: number | ((attempt: number) => number);
  /**
   * Request type.
   * @default: 'json'
   */
  type?: 'json' | 'urlencoded';
  url: string;
  /**
   * Wait for the user to trigger the request.
   * @default: false
   */
  wait?: boolean;
}

interface UseFetchError extends Error {
  response?: unknown;
  status?: number;
}

interface UseFetchState<TDataType> {
  data?: TDataType;
  error?: UseFetchError;
  isCached: boolean;
  status: UseFetchStatus;
  url: string;
}

interface UseFetchResult<TDataType> extends UseFetchState<TDataType> {
  /**
   * Whether the response is cached.
   */
  isCached: boolean;
  isError: () => boolean;
  isFetched: () => boolean;
  isLoading: () => boolean;
  isPaused: () => boolean;
  isSuccess: () => boolean;
  refetch: (eraseData?: boolean) => void;
}

interface UseFetchCache {
  /**
   * Retrieves cached data for the given URL if available and not expired.
   * @param url The URL to retrieve data for.
   * @returns The cached data, or `undefined` if no valid cache exists.
   */
  get: (url: string) => any | undefined;

  /**
   * Caches data for the given URL with a specified time-to-live (TTL).
   * @param url The URL to cache data for.
   * @param data The data to cache.
   * @param ttl Time-to-live for the cached data in milliseconds.
   */
  set: (url: string, data: any, ttl: number) => void;

  /**
   * Clears cached data. If a URL is provided, only the cache for that URL is cleared.
   * If no URL is specified, the entire cache is cleared.
   * @param url Optional URL to clear the cache for.
   */
  clear: (url?: string) => void;

  /**
   * Checks if valid cached data exists for the given URL.
   * @param url The URL to check.
   * @returns `true` if the URL is cached and still valid, `false` otherwise.
   */
  has: (url: string) => boolean;
}

useFetch<TDataType = unknown>(urlOrOptions: string | UseFetchOptions): UseFetchResult<TDataType>;
```
