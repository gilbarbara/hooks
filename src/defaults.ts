import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
