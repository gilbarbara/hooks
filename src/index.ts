export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
