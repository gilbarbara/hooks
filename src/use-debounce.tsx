import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
