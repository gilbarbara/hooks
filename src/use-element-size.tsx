import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
