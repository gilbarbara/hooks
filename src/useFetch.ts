import { useCallback, useEffect, useMemo } from 'react';

import { PlainObject } from './types';
import { useIsMounted } from './useIsMounted';
import { useMemoizedValue } from './useMemoizedValue';
import { usePrevious } from './usePrevious';
import { useSetState } from './useSetState';
import { isPlainObject, isURL } from './utils';

export const USE_FETCH_STATUS = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
} as const;

interface UseFetchError extends Error {
  response?: unknown;
  status?: number;
}

interface UseFetchState<TDataType> {
  data?: TDataType;
  error?: UseFetchError;
  isCached: boolean;
  retryCount: number | null;
  status: UseFetchStatus;
  url: string;
}

export type UseFetchStatus = keyof typeof USE_FETCH_STATUS;

export interface UseFetchOptions {
  body?: BodyInit | Record<string, any>;
  /**
   * Time to cache the request if provided.
   * When the cache is expired, the request will be triggered again.
   */
  cacheTTL?: number;
  headers?: PlainObject<string>;
  /**
   * HTTP method.
   * @default: 'GET'
   */
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'HEAD' | 'OPTIONS' | 'CONNECT' | 'TRACE';
  /**
   * Request mode.
   * @default: 'cors'
   */
  mode?: 'cors' | 'navigate' | 'no-cors' | 'same-origin';
  /** Callback fired when an error occurs */
  onError?: (error: UseFetchError) => void;
  /** Callback fired when the request completes (success or error) */
  onFinally?: () => void;
  /** Callback fired when the loading state changes */
  onLoading?: () => void;
  /** Callback fired when data is successfully fetched */
  onSuccess?: (data: any) => void;
  /**
   * Number of retries.
   */
  retries?: number;
  /**
   * Time to wait before retrying.
   * A function like attempt => Math.min(attempt > 1 ? 2 ** attempt * 1000 : 1000, 30 * 1000) applies exponential backoff.
   * A function like attempt => attempt * 1000 applies linear backoff.
   * @default: attempt => attempt * 1000
   */
  retryDelay?: number | ((attempt: number) => number);
  /**
   * Request type.
   * @default: 'json'
   */
  type?: 'json' | 'urlencoded';
  url: string;
  /**
   * Wait for the user to trigger the request.
   * @default: false
   */
  wait?: boolean;
}

export interface UseFetchResult<TDataType> extends UseFetchState<TDataType> {
  /**
   * Whether the response is cached.
   */
  isCached: boolean;
  isError: () => boolean;
  isFetched: () => boolean;
  isLoading: () => boolean;
  isPaused: () => boolean;
  isSuccess: () => boolean;
  refetch: (eraseData?: boolean) => void;
}

const globalCache = new Map<string, { data: any; expiry: number }>();

async function request(options: UseFetchOptions): Promise<any> {
  const {
    body = undefined,
    headers = {},
    method = 'GET',
    mode = 'cors',
    type = 'json',
    url = '',
  } = options;

  const contentTypes = {
    json: 'application/json',
    urlencoded: 'application/x-www-form-urlencoded',
  };

  const params: RequestInit = {
    body: undefined,
    cache: 'no-store' as const,
    headers: {
      Accept: 'application/json',
      'Content-Type': contentTypes[type],
      ...headers,
    },
    method,
    mode,
    credentials: undefined,
  };

  if (body) {
    params.body = (
      isPlainObject(body) && type === 'json' ? JSON.stringify(body) : body
    ) as BodyInit;
  }

  try {
    const response = await fetch(url, params);
    let content: unknown;

    try {
      content = await response.json();
    } catch {
      content = await response.text();
    }

    if (response.status > 299) {
      const error = new Error(response.statusText) as UseFetchError;

      error.status = response.status;
      error.response = content;

      throw error;
    }

    return content;
  } catch (error) {
    // Handle hard failures (network issues, CORS errors, fetch rejection)
    const fetchError = new Error('Network request failed') as UseFetchError;

    fetchError.status = 0; // Status 0 indicates a network error
    fetchError.response = error instanceof Error ? error.message : error;
    throw fetchError;
  }
}

export const useFetchCache = {
  /**
   * Retrieve data from the cache.
   * @param url The URL to retrieve data for.
   * @returns Cached data or `undefined` if not found or expired.
   */
  get(url: string) {
    const cached = globalCache.get(url);

    if (cached && Date.now() < cached.expiry) {
      return cached;
    }

    globalCache.delete(url); // Remove expired entry

    return undefined;
  },

  /**
   * Set data in the cache.
   * @param url The URL to cache data for.
   * @param data The data to cache.
   * @param ttl Time-to-live in milliseconds.
   */
  set(url: string, data: any, ttl: number) {
    globalCache.set(url, { data, expiry: Date.now() + ttl });
  },

  /**
   * Clear the cache.
   * @param url Optional URL to clear. Clears all cache if not specified.
   */
  clear(url?: string) {
    if (url) {
      globalCache.delete(url);
    } else {
      globalCache.clear();
    }
  },

  /**
   * Check if a URL is cached and still valid.
   * @param url The URL to check.
   * @returns `true` if cached and valid, `false` otherwise.
   */
  has(url: string) {
    const cached = globalCache.get(url);

    return !!cached && Date.now() < cached.expiry;
  },
};

export function useFetch<TDataType = unknown>(
  urlOrOptions: string | UseFetchOptions,
): UseFetchResult<TDataType> {
  const {
    cacheTTL = 0,
    onError,
    onFinally,
    onLoading,
    onSuccess,
    retries = 0,
    retryDelay = (attempt: number) => attempt * 1000,
    wait = false,
    ...options
  } = useMemoizedValue(
    (isURL(urlOrOptions)
      ? ({ type: 'json', url: urlOrOptions } as const)
      : urlOrOptions) as UseFetchOptions,
  );
  const [{ data, error, isCached, retryCount, status, url }, setState] = useSetState<
    UseFetchState<TDataType>
  >({
    data: undefined,
    error: undefined,
    isCached: false,
    retryCount: null,
    status: USE_FETCH_STATUS.IDLE,
    url: options.url,
  });
  const isMounted = useIsMounted();
  const previousRetryCount = usePrevious(retryCount);

  if (!isPlainObject(options) || !isURL(url)) {
    throw new Error('Expected an options object or URL');
  }

  const getData = useCallback(
    (eraseData?: boolean) => {
      setState(s => ({
        data: eraseData ? undefined : s.data,
        error: undefined,
        isCached: false,
        status: USE_FETCH_STATUS.LOADING,
      }));

      onLoading?.();

      // Check cache
      if (cacheTTL && useFetchCache.has(url) && !eraseData) {
        const cached = useFetchCache.get(url);

        if (cached) {
          setState({
            data: cached.data,
            error: undefined,
            isCached: true,
            status: USE_FETCH_STATUS.SUCCESS,
          });

          onSuccess?.(cached.data);
          onFinally?.();

          return;
        }
      }

      request({ ...options })
        .then(response => {
          if (!isMounted()) {
            return;
          }

          if (cacheTTL) {
            useFetchCache.set(url, response, cacheTTL);
          }

          setState({
            data: response,
            retryCount: null,
            status: USE_FETCH_STATUS.SUCCESS,
          });
          onSuccess?.(response);
        })
        .catch(responseError => {
          if (!isMounted()) {
            return;
          }

          const counter = retryCount ?? 0;

          if (!retries || counter >= retries) {
            setState({
              error: responseError,
              status: USE_FETCH_STATUS.ERROR,
            });
            onError?.(responseError);
          }

          if (retries && counter < retries) {
            setState({
              retryCount: counter + 1,
            });
          }
        })
        .finally(() => {
          if (isMounted()) {
            onFinally?.();
          }
        });
    },
    [
      cacheTTL,
      isMounted,
      onError,
      onFinally,
      onLoading,
      onSuccess,
      options,
      retries,
      retryCount,
      setState,
      url,
    ],
  );

  useEffect(() => {
    if (url !== options.url) {
      setState({
        data: undefined,
        error: undefined,
        status: USE_FETCH_STATUS.IDLE,
        url: options.url,
      });
    }
  }, [options.url, setState, url]);

  useEffect(() => {
    if (status === USE_FETCH_STATUS.IDLE && !wait) {
      getData();
    }
  }, [getData, status, wait]);

  useEffect(() => {
    if (retries && typeof retryCount === 'number' && retryCount !== previousRetryCount) {
      setTimeout(getData, typeof retryDelay === 'function' ? retryDelay(retryCount) : retryDelay);
    }
  }, [getData, previousRetryCount, retries, retryCount, retryDelay]);

  const isError = useCallback(() => status === USE_FETCH_STATUS.ERROR, [status]);
  const isFetched = useCallback(
    () => ([USE_FETCH_STATUS.SUCCESS, USE_FETCH_STATUS.ERROR] as UseFetchStatus[]).includes(status),
    [status],
  );
  const isLoading = useCallback(() => status === USE_FETCH_STATUS.LOADING, [status]);
  const isPaused = useCallback(() => status === USE_FETCH_STATUS.IDLE && wait, [status, wait]);
  const isSuccess = useCallback(() => status === USE_FETCH_STATUS.SUCCESS, [status]);
  const refetch = useCallback((eraseData = false) => getData(eraseData), [getData]);

  return useMemo(
    () => ({
      data,
      error,
      isCached,
      isError,
      isFetched,
      isLoading,
      isPaused,
      isSuccess,
      refetch,
      retryCount,
      status,
      url,
    }),
    [
      data,
      error,
      isCached,
      isError,
      isFetched,
      isLoading,
      isPaused,
      isSuccess,
      refetch,
      retryCount,
      status,
      url,
    ],
  );
}
