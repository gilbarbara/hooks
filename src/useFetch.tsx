import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

import { PlainObject } from './types';
import { isPlainObject, isURL } from './utils';

export const USE_FETCH_STATUS = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
} as const;

export type UseFetchStatus = keyof typeof USE_FETCH_STATUS;

interface UseFetchError extends Error {
  response?: unknown;
  status?: number;
}

export interface UseFetchOptions {
  body?: BodyInit | Record<string, any>;
  headers?: PlainObject<string>;
  /**
   * HTTP method.
   * @default: 'GET'
   */
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'HEAD' | 'OPTIONS' | 'CONNECT' | 'TRACE';
  /**
   * Request mode.
   * @default: 'cors'
   */
  mode?: 'cors' | 'navigate' | 'no-cors' | 'same-origin';
  /**
   * Number of retries.
   */
  retry?: number;
  /**
   * Time to wait before retrying.
   * A function like attempt => Math.min(attempt > 1 ? 2 ** attempt * 1000 : 1000, 30 * 1000) applies exponential backoff.
   * A function like attempt => attempt * 1000 applies linear backoff.
   * @default: attempt => attempt * 1000
   */
  retryDelay?: number | ((attempt: number) => number);
  /**
   * Request type.
   * @default: 'json'
   */
  type?: 'json' | 'urlencoded';
  url: string;
  /**
   * Wait for the user to trigger the request.
   * @default: false
   */
  wait?: boolean;
}

interface UseFetchState<TDataType> {
  data?: TDataType;
  error?: UseFetchError;
  status: UseFetchStatus;
}

export interface UseFetchResult<TDataType> extends UseFetchState<TDataType> {
  isError: () => boolean;
  isFetched: () => boolean;
  isLoading: () => boolean;
  isPaused: () => boolean;
  isSuccess: () => boolean;
  refetch: (eraseData?: boolean) => void;
}

async function request(options: UseFetchOptions): Promise<any> {
  const {
    body = undefined,
    headers = {},
    method = 'GET',
    mode = 'cors',
    type = 'json',
    url = '',
  } = options;

  const contentTypes = {
    json: 'application/json',
    urlencoded: 'application/x-www-form-urlencoded',
  };

  const params: RequestInit = {
    body: undefined,
    cache: 'no-store' as const,
    headers: {
      Accept: 'application/json',
      'Content-Type': contentTypes[type],
      ...headers,
    },
    method,
    mode,
    credentials: undefined,
  };

  if (body) {
    params.body = (
      isPlainObject(body) && type === 'json' ? JSON.stringify(body) : body
    ) as BodyInit;
  }

  return fetch(url, params).then(async response => {
    const text = await response.text();
    let content: unknown;

    try {
      content = JSON.parse(text);
    } catch {
      content = text;
    }

    if (response.status > 299) {
      const error = new Error(response.statusText) as UseFetchError;

      error.status = response.status;
      error.response = content;

      throw error;
    }

    return content;
  });
}

export function useFetch<TDataType = unknown>(
  urlOrOptions: string | UseFetchOptions,
): UseFetchResult<TDataType> {
  const isActive = useRef(false);
  const retryCount = useRef(0);
  const [{ data, error, status }, setState] = useState<UseFetchState<TDataType>>({
    data: undefined,
    error: undefined,
    status: USE_FETCH_STATUS.IDLE,
  });
  const {
    retry = 0,
    retryDelay = (attempt: number) => attempt * 1000,
    wait = false,
    ...options
  } = isURL(urlOrOptions)
    ? ({
        type: 'json',
        url: urlOrOptions,
      } as const)
    : urlOrOptions;

  if (!isPlainObject(options) || !isURL(options.url)) {
    throw new Error('Expected an options object or URL');
  }

  const getData = useCallback(
    (eraseData?: boolean) => {
      setState(s => ({
        ...s,
        data: eraseData ? undefined : s.data,
        error: undefined,
        status: USE_FETCH_STATUS.LOADING,
      }));

      request(options)
        .then(response => {
          if (isActive.current) {
            setState(s => ({
              ...s,
              data: response,
              status: USE_FETCH_STATUS.SUCCESS,
            }));
          }
        })
        .catch(responseError => {
          if (isActive.current) {
            setState(s => ({
              ...s,
              error: responseError,
              status: USE_FETCH_STATUS.ERROR,
            }));
          }

          if (retry && retryCount.current < retry) {
            retryCount.current += 1;

            setTimeout(
              getData,
              typeof retryDelay === 'function' ? retryDelay(retryCount.current) : retryDelay,
            );
          }
        });
    },
    [options, retry, retryDelay],
  );

  useEffect(() => {
    isActive.current = true;

    return () => {
      isActive.current = false;
    };
  }, []);

  useEffect(() => {
    if (status === USE_FETCH_STATUS.IDLE && !wait) {
      getData();
    }
  }, [getData, status, wait]);

  const isError = useCallback(() => status === USE_FETCH_STATUS.ERROR, [status]);

  const isFetched = useCallback(
    () => ([USE_FETCH_STATUS.SUCCESS, USE_FETCH_STATUS.ERROR] as UseFetchStatus[]).includes(status),
    [status],
  );

  const isLoading = useCallback(() => status === USE_FETCH_STATUS.LOADING, [status]);

  const isPaused = useCallback(() => status === USE_FETCH_STATUS.IDLE && wait, [status, wait]);

  const isSuccess = useCallback(() => status === USE_FETCH_STATUS.SUCCESS, [status]);
  const refetch = useCallback((eraseData = false) => getData(eraseData), [getData]);

  return useMemo(
    () => ({
      data,
      error,
      isError,
      isFetched,
      isLoading,
      isPaused,
      isSuccess,
      refetch,
      status,
    }),
    [data, error, isError, isFetched, isLoading, isPaused, isSuccess, refetch, status],
  );
}
