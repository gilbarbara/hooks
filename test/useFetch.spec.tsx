import * as React from 'react';
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';

import repositories from './__fixtures__/repositories.json';

import { PlainObject } from '../src/types';
import { useFetch } from '../src/useFetch';

const url = 'https://api.github.com/search/repositories?q=react&sort=stars';

function Component({ fetchOptions, skip }: any) {
  const { data, error, status } = useFetch<PlainObject<any>[]>(fetchOptions, skip);

  return (
    <div data-testid="content">
      {status === 'failure' && <p data-testid="failure">{error?.toString()}</p>}
      {status === 'success' && (
        <ul data-testid="success">
          {data?.map(d => (
            <li key={d.id}>{d.full_name}</li>
          ))}
        </ul>
      )}
      {status === 'running' && <p data-testid="running">Loading</p>}
    </div>
  );
}

describe('useFetch', () => {
  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => undefined);
  });

  afterAll(() => {
    // @ts-ignore
    console.error.mockRestore();
  });

  afterEach(() => {
    fetchMock.mockClear();
  });

  it('should handle success', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(repositories));

    const { rerender } = render(<Component fetchOptions={url} />);

    expect(fetchMock).toHaveBeenCalledWith(
      'https://api.github.com/search/repositories?q=react&sort=stars',
      {
        body: undefined,
        cache: 'no-store',
        credentials: undefined,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        method: 'GET',
        mode: 'cors',
      },
    );

    expect(screen.getByTestId('running')).toMatchSnapshot();

    rerender(<Component fetchOptions={url} />);

    await waitForElementToBeRemoved(() => screen.queryByTestId('running'));

    expect(screen.getByTestId('success')).toMatchSnapshot();
  });

  it('should handle hard failure', async () => {
    fetchMock.mockRejectOnce(new Error('Failed to fetch'));

    render(
      <Component
        fetchOptions={{
          url,
          type: 'urlencoded',
        }}
      />,
    );

    expect(fetchMock).toHaveBeenCalledWith(
      'https://api.github.com/search/repositories?q=react&sort=stars',
      {
        body: undefined,
        cache: 'no-store',
        credentials: undefined,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        method: 'GET',
        mode: 'cors',
      },
    );

    await waitForElementToBeRemoved(() => screen.queryByTestId('running'));

    expect(screen.getByTestId('failure')).toMatchSnapshot();
  });

  it('should handle soft failure', async () => {
    fetchMock.mockResponse('Request failed', {
      status: 400,
      statusText: 'Request failed',
    });

    render(
      <Component
        fetchOptions={{
          url,
          method: 'POST',
          body: { a: 1 },
          headers: {
            Authorization: 'Bearer 263619d6-1da5-41fc-ba66-081acacbf9af',
          },
        }}
      />,
    );

    expect(fetchMock).toHaveBeenCalledWith(
      'https://api.github.com/search/repositories?q=react&sort=stars',
      {
        body: '{"a":1}',
        cache: 'no-store',
        credentials: undefined,
        headers: {
          Accept: 'application/json',
          Authorization: 'Bearer 263619d6-1da5-41fc-ba66-081acacbf9af',
          'Content-Type': 'application/json',
        },
        method: 'POST',
        mode: 'cors',
      },
    );

    await waitForElementToBeRemoved(() => screen.queryByTestId('running'));

    expect(screen.getByTestId('failure')).toMatchSnapshot();
  });

  it('should handle "skip"', async () => {
    render(<Component fetchOptions={url} skip />);

    expect(fetchMock).toHaveBeenCalledTimes(0);

    expect(screen.getByTestId('content')).toMatchSnapshot();
  });

  it('should handle throw with invalid parameters', async () => {
    expect(() => render(<Component fetchOptions={[]} />)).toThrow(
      'Expected an options object or URL',
    );
  });
});
