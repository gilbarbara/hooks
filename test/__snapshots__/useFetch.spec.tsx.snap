// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFetch should handle "skip" 1`] = `
<div
  data-testid="content"
/>
`;

exports[`useFetch should handle hard failure 1`] = `
<p
  data-testid="failure"
>
  Error: Failed to fetch
</p>
`;

exports[`useFetch should handle soft failure 1`] = `
<p
  data-testid="failure"
>
  Error: Request failed
</p>
`;

exports[`useFetch should handle success 1`] = `
<p
  data-testid="running"
>
  Loading
</p>
`;

exports[`useFetch should handle success 2`] = `
<ul
  data-testid="success"
>
  <li>
    mojombo/grit
  </li>
  <li>
    wycats/merb-core
  </li>
  <li>
    rubinius/rubinius
  </li>
  <li>
    mojombo/god
  </li>
  <li>
    vanpelt/jsawesome
  </li>
  <li>
    wycats/jspec
  </li>
  <li>
    defunkt/exception_logger
  </li>
  <li>
    defunkt/ambition
  </li>
  <li>
    technoweenie/restful-authentication
  </li>
  <li>
    technoweenie/attachment_fu
  </li>
</ul>
`;
